src/backend/distributed/planner/merge_planner.c

```diff
@@ -554,27 +669,335 @@ MergeQualAndTargetListFunctionsSupported(Oid resultRelationId, FromExpr *joinTre
 }
 
 
-#endif
+/*
+ * RepartitionMergeSupported checks if certain conditions cannot accommodate the
+ * strategy of repartition and redistribution of source rows, the routine will verify
+ * them and subsequently raises an exception.
+ */
+static void
+ErrorIfRepartitionMergeNotSupported(Oid targetRelationId, Query *mergeQuery,
+									Query *sourceQuery)
+{
+	if (!IsCitusTableType(targetRelationId, DISTRIBUTED_TABLE))
+	{
+		ereport(ERROR,
+				(errmsg("MERGE involving repartition of rows "
+						"is supported only if the target is distributed")));
+	}
+
+	RTEListProperties *queryRteListProperties = GetRTEListPropertiesForQuery(mergeQuery);
+	if (queryRteListProperties->hasPostgresLocalTable)
+	{
+		ereport(ERROR, (errmsg("MERGE INTO an distributed table from "
+							   "Postgres table is not yet supported")));
+	}
+
+	queryRteListProperties = GetRTEListPropertiesForQuery(sourceQuery);
+	if (!queryRteListProperties->hasCitusTable)
+	{
+		ereport(ERROR, (errmsg("To MERGE into a distributed table, source must "
+							   "be Citus table(s)")));
+	}
+}
 
 
 /*
- * MergeQuerySupported does check for a MERGE command in the query, if it finds
- * one, it will verify the below criteria
- * - Supported tables and combinations in ErrorIfMergeHasUnsupportedTables
- * - Distributed tables requirements in ErrorIfDistTablesNotColocated
- * - Checks target-lists and functions-in-quals in TargetlistAndFunctionsSupported
+ * ConvertCteRTEIntoSubquery takes a RTE_CTE and converts it into a RTE_SUBQUERY.
  */
-DeferredErrorMessage *
-MergeQuerySupported(Oid resultRelationId, Query *originalQuery, bool multiShardQuery,
-					PlannerRestrictionContext *plannerRestrictionContext)
+static void
+ConvertCteRTEIntoSubquery(Query *mergeQuery, RangeTblEntry *sourceRte)
 {
-	/* function is void for pre-15 versions of Postgres */
-	#if PG_VERSION_NUM < PG_VERSION_15
+	CommonTableExpr *sourceCte = NULL;
+	CommonTableExpr *candidateCte = NULL;
+	List *cteList = NIL;
```

Comment:
I think we should add a comment here stating that this logic works because in practice we only allow 1 cte to be used in the MERGE command itself, only on the USING clause.

If we were to allow ctes in MergeCondition, this would clearly not work. We cannot really inline all CTEs into the USING source.

This currently works because the other CTEs in `sourceCte` are either used in `sourceCte` or not used at all.

I think the same applies for the ctes in subquery conversion. If the Merge->cteList is allowed to be used in other parts of the query, that logic would clearly be broken

Reply:
Added the comments for brevity


 ```
        /*
         * Presently, CTEs are only permitted within the USING clause, so we
         * search for the matching one
         */

```